packages/puppeteer-core/src/injected/PSelectorTokenizer.ts

```diff
@@ -0,0 +1,251 @@
+import {assert} from '../util/assert.js';
+
+export const enum TokenType {
+  Class = 'class',
+  Attribute = 'attribute',
+  Id = 'id',
+  Type = 'type',
+  Universal = 'universal',
+  PseudoElement = 'pseudo-element',
+  PseudoClass = 'pseudo-class',
+  Comma = 'comma',
+  Combinator = 'combinator',
+}
+
+export interface Token {
+  type: string;
+  content: string;
+  name: string;
+  namespace?: string;
+  value?: string;
+  pos: [number, number];
+  operator?: string;
+  argument?: string;
+  caseSensitive?: 'i';
+  /**
+   * @internal
+   */
+  __changed?: boolean;
+}
+
+const TOKENS: Record<string, RegExp> = {
+  [TokenType.Attribute]:
+    /\[\s*(?:(?<namespace>(?:\\.|[-\w\P{ASCII}])+|\*)?\|)?(?<name>(?:\\.|[-\w\P{ASCII}])+)\s*(?:(?<operator>\W?=)\s*(?<value>.+?)\s*(\s(?<caseSensitive>[iIsS]))?\s*)?\]/gu,
+  [TokenType.Id]: /#(?<name>(?:\\.|[-\w\P{ASCII}])+)/gu,
+  [TokenType.Class]: /\.(?<name>(?:\\.|[-\w\P{ASCII}])+)/gu,
+  [TokenType.Comma]: /\s*,\s*/g,
+  [TokenType.Combinator]: /\s*(?:>{3,4}|[\s>+~])\s*/g,
+  [TokenType.PseudoElement]:
+    /::(?<name>(?:\\.|[-\w\P{ASCII}])+)(?:\((?<argument>¶+)\))?/gu,
+  [TokenType.PseudoClass]:
+    /:(?<name>(?:\\.|[-\w\P{ASCII}])+)(?:\((?<argument>¶+)\))?/gu,
+  [TokenType.Universal]: /(?:(?<namespace>\*|(?:\\.|[-\w\P{ASCII}])*)\|)?\*/gu,
+  [TokenType.Type]:
+    /(?:(?<namespace>\*|(?:\\.|[-\w\P{ASCII}])*)\|)?(?<name>(?:\\.|[-\w\P{ASCII}])+)/gu,
+};
+
+const getArgumentPatternByType = (type: string) => {
+  switch (type) {
+    case TokenType.PseudoElement:
+    case TokenType.PseudoClass:
+      return new RegExp(
+        TOKENS[type]!.source.replace('(?<argument>¶+)', '(?<argument>.+)'),
+        'gu'
+      );
+    default:
+      return TOKENS[type];
+  }
+};
+
+function assertTokenArray(
+  tokens: Array<Token | string>
+): asserts tokens is Token[] {
+  let offset = 0;
+  for (const token of tokens) {
+    switch (typeof token) {
+      case 'string':
+        throw new Error(
+          `Unexpected sequence ${token} found at index ${offset}`
+        );
+      case 'object':
+        offset += token.content.length;
+        token.pos = [offset - token.content.length, offset];
+        switch (token.type) {
+          case TokenType.Combinator:
+          case TokenType.Comma:
+            token.content = token.content.trim() || ' ';
+            break;
+        }
+        break;
+    }
+  }
+}
+
+export function tokenize(selector: string, grammar = TOKENS): Token[] {
```

Comment:
Include tests for the tokenize functions with most common cases

Reply:
The tests for P selectors should sufficiently cover the tokenizer.